File: server/src/main/java/com/krillsson/sysapi/core/monitoring/monitors/CpuMonitor.java
package com.krillsson.sysapi.core.monitoring.monitors;

import com.krillsson.sysapi.core.domain.system.SystemLoad;
import com.krillsson.sysapi.core.monitoring.Monitor;
import com.krillsson.sysapi.core.monitoring.MonitorType;

import java.time.Duration;

public class CpuMonitor extends Monitor {

    public CpuMonitor(String id, Duration inertia, double threshold) {
        super(id, inertia, threshold);
    }

    @Override
    protected double value(SystemLoad systemLoad) {
        return systemLoad.getCpuLoad().getCpuLoadOsMxBean();
    }

    @Override
    protected boolean isOutsideThreshold(double value) {
        return value > threshold();
    }

    @Override
    protected MonitorType type() {
        return MonitorType.CPU_LOAD;
    }

    @Override
    public String toString() {
        return "CpuMonitor{" +
                "threshold=" + threshold() +
                '}';
    }
}

File: server/src/main/java/com/krillsson/sysapi/core/monitoring/monitors/CpuMonitor.kt
package com.krillsson.sysapi.core.monitoring.monitors

import com.krillsson.sysapi.core.domain.monitor.MonitorConfig
import com.krillsson.sysapi.core.domain.monitor.MonitoredValue
import com.krillsson.sysapi.core.monitoring.MetricQueryEvent
import com.krillsson.sysapi.core.monitoring.Monitor
import java.util.*

class CpuMonitor(override val id: UUID, override val config: MonitorConfig<MonitoredValue.FractionalValue>) :
    Monitor<MonitoredValue.FractionalValue>() {

    companion object {
        val selector: FractionalValueSelector = { load, _ ->
            MonitoredValue.FractionalValue(load.cpuLoad.usagePercentage.toFloat())
        }
    }

    override val type: Type = Type.CPU_LOAD

    override fun selectValue(event: MetricQueryEvent): MonitoredValue.FractionalValue? =
        selector(event.load, null)

    override fun isPastThreshold(value: MonitoredValue.FractionalValue): Boolean {
        return value > config.threshold
    }
}